packages/modeling/src/operations/modifiers/insertTjunctions.js

```diff
@@ -0,0 +1,294 @@
+    sidemap.set(newsidetag, [newsideobj])
+  } else {
+    sidemap.get(newsidetag).push(newsideobj)
+  }
+  if (starttag in vertextag2sidestart) {
+    vertextag2sidestart[starttag].push(newsidetag)
+  } else {
+    vertextag2sidestart[starttag] = [newsidetag]
+  }
+  if (endtag in vertextag2sideend) {
+    vertextag2sideend[endtag].push(newsidetag)
+  } else {
+    vertextag2sideend[endtag] = [newsidetag]
+  }
+  return newsidetag
+}
+
+const deleteSide = (sidemap, vertextag2sidestart, vertextag2sideend, vertex0, vertex1, polygonindex) => {
+  const starttag = getTag(vertex0)
+  const endtag = getTag(vertex1)
+  const sidetag = `${starttag}/${endtag}`
+  if (assert && !(sidemap.has(sidetag))) throw new Error('assert failed')
+  let idx = -1
+  const sideobjs = sidemap.get(sidetag)
+  for (let i = 0; i < sideobjs.length; i++) {
+    const sideobj = sideobjs[i]
+    let sidetag = getTag(sideobj.vertex0)
+    if (sidetag !== starttag) continue
+    sidetag = getTag(sideobj.vertex1)
+    if (sidetag !== endtag) continue
+    if (polygonindex !== null) {
+      if (sideobj.polygonindex !== polygonindex) continue
+    }
+    idx = i
+    break
+  }
+  if (assert && idx < 0) throw new Error('assert failed')
+  sideobjs.splice(idx, 1)
+  if (sideobjs.length === 0) {
+    sidemap.delete(sidetag)
+  }
+
+  // adjust start and end lists
+  idx = vertextag2sidestart[starttag].indexOf(sidetag)
+  if (assert && idx < 0) throw new Error('assert failed')
+  vertextag2sidestart[starttag].splice(idx, 1)
+  if (vertextag2sidestart[starttag].length === 0) {
+    delete vertextag2sidestart[starttag]
+  }
+
+  idx = vertextag2sideend[endtag].indexOf(sidetag)
+  if (assert && idx < 0) throw new Error('assert failed')
+  vertextag2sideend[endtag].splice(idx, 1)
+  if (vertextag2sideend[endtag].length === 0) {
+    delete vertextag2sideend[endtag]
+  }
+}
+
+/*
+  Suppose we have two polygons ACDB and EDGF:
+
+   A-----B
+   |     |
+   |     E--F
+   |     |  |
+   C-----D--G
+
+  Note that vertex E forms a T-junction on the side BD. In this case some STL slicers will complain
+  that the solid is not watertight. This is because the watertightness check is done by checking if
+  each side DE is matched by another side ED.
+
+  This function will return a new solid with ACDB replaced by ACDEB
+
+  Note that this can create polygons that are slightly non-convex (due to rounding errors). Therefore the result should
+  not be used for further CSG operations!
+
+  Note this function is meant to be used to preprocess geometries when triangulation is required, i.e. AMF, STL, etc.
+  Do not use the results in other operations.
+*/
+
+/*
+ * Insert missing vertices for T junctions, which creates polygons that can be triangulated.
+ * @param {Array} polygons - the original polygons which may or may not have T junctions
+ * @return original polygons (if no T junctions found) or new polygons with updated vertices
+ */
+const insertTjunctions = (polygons) => {
```

Comment:
This method is a bit enormous. Are there any easily extractable chunks we could pull out to make it more readable/maintainable?

Reply:
Actually, there are three sets of 'edge' implementations in these functions.
- edges.js
- edges in insertTjunctions.js
- edges in mergePolygons.js

A single implement would be ideal but I'm not really motivated. I'm kind of tired of modifiers.

Reply:
Oh. And there's another in the quickhull piece as well.

Reply:
@SimonClark thoughts? how about a new issue to revisit this later?

Reply:
sounds fine